src/main/scala/com/databricks/labs/overwatch/pipeline/GoldTransforms.scala

```diff
@@ -265,58 +319,35 @@ trait GoldTransforms extends SparkSessionWrapper {
       .select(
-        'organization_id, 'cluster_id,
-        last('type, true).over(stateUnboundW).alias("last_state_previous_run")
+        'organization_id, 'cluster_id, 'isRunning,
+        'timestamp, 'type, 'current_num_workers, 'target_num_workers
       )
 
-    // don't allow cluster to change state from running to running or vice-versa. When this occurs just set cluster
-    // to terminated and reset / restart.
-    val invalidEventChain = lead('runningSwitch, 1).over(stateUnboundW).isNotNull &&
-      lead('runningSwitch, 1).over(stateUnboundW) === lead('previousSwitch, 1).over(stateUnboundW)
-    val clusterEventsBaseline = clusterEventsDF
-      .selectExpr("*", "details.*")
-      .drop("details")
-      .join(clusterStateBeforeRun, Seq("organization_id", "cluster_id"), "left")
-      .withColumn( // if incremental df doesn't have a running switch get it from latest previous state
-        "runningSwitch",
-        coalesce(getRunningSwitch('type), getRunningSwitch('last_state_previous_run))
-      )
-      .withColumn(
-        "previousSwitch",
-        when('runningSwitch.isNotNull, last('runningSwitch, true).over(stateUntilPreviousRowW))
-      )
-      .withColumn(
-        "invalidEventChainHandler",
-        when(invalidEventChain, array(lit(false), lit(true))).otherwise(array(lit(false)))
-      )
-      .selectExpr("*", "explode(invalidEventChainHandler) as imputedTerminationEvent").drop("invalidEventChainHandler")
-      .withColumn("type", when('imputedTerminationEvent, "TERMINATING").otherwise('type))
-      .withColumn("timestamp", when('imputedTerminationEvent, lag('timestamp, 1).over(stateUnboundW) + 1L).otherwise('timestamp))
-      .withColumn("isRunning", when('imputedTerminationEvent, lit(false)).otherwise(last('runningSwitch, true).over(stateUntilCurrentW)))
-      .withColumn(
-        "current_num_workers",
-        when(!'isRunning || 'isRunning.isNull, lit(null).cast("long"))
-          .otherwise(coalesce('current_num_workers, $"cluster_size.num_workers", $"cluster_size.autoscale.min_workers", last(coalesce('current_num_workers, $"cluster_size.num_workers", $"cluster_size.autoscale.min_workers"), true).over(stateUntilCurrentW)))
-      )
-      .withColumn(
-        "target_num_workers",
-        when(!'isRunning || 'isRunning.isNull, lit(null).cast("long"))
-          .when('type === "CREATING", coalesce($"cluster_size.num_workers", $"cluster_size.autoscale.min_workers"))
-          .otherwise(coalesce('target_num_workers, 'current_num_workers))
-      )
+    val clusterEventsForCurrentRun = getEventStates(clusterEventsDF, 'timestamp >= fromTime.asUnixTimeMilli)
       .select(
         'organization_id, 'cluster_id, 'isRunning,
         'timestamp, 'type, 'current_num_workers, 'target_num_workers
       )
 
+    val allClusterEventsInScope = clusterStateBeforeRun.unionByName(clusterEventsForCurrentRun)
```

Comment:
generates the state of the event before the current run and pulls it forward to properly close the event's end time. This is currently necessary because we aren't yet doing merges. Merging DFs in next version will greatly simplify and improve this entire pipe.